art/attacks/evasion/dpatch.py

```diff
@@ -0,0 +1,261 @@
+
+            patch_gradients = np.zeros_like(self._patch)
+
+            for i_batch in range(num_batches):
+                i_batch_start = i_batch * self.batch_size
+                i_batch_end = min((i_batch + 1) * self.batch_size, patched_images.shape[0])
+
+                gradients = self.estimator.loss_gradient(
+                    x=patched_images[i_batch_start:i_batch_end], y=patch_target[i_batch_start:i_batch_end]
+                )
+
+                for i_image in range(self.batch_size):
+
+                    i_x_1 = transforms[i_batch_start + i_image]["i_x_1"]
+                    i_x_2 = transforms[i_batch_start + i_image]["i_x_2"]
+                    i_y_1 = transforms[i_batch_start + i_image]["i_y_1"]
+                    i_y_2 = transforms[i_batch_start + i_image]["i_y_2"]
+
+                    if self.estimator.channel_index == 3:
+                        patch_gradients_i = gradients[i_image, i_x_1:i_x_2, i_y_1:i_y_2, :]
+                    elif self.estimator.channel_index == 1:
+                        patch_gradients_i = gradients[i_image, :, i_x_1:i_x_2, i_y_1:i_y_2]
+                    else:
+                        raise ValueError("Unrecognized channel index.")
+
+                    patch_gradients += patch_gradients_i
+
+            self._patch -= patch_gradients * self.learning_rate
+            self._patch = np.clip(self._patch, a_min=self.estimator.clip_values[0], a_max=self.estimator.clip_values[1])
+
+        return self._patch
+
+    @staticmethod
+    def _augment_images_with_patch(x, patch, random_location, channel_index):
+        """
+        Augment images with randomly shifted patch.
+        :param x: Sample images.
+        :type x: `np.ndarray`
+        :param patch: The patch to be applied.
+        :param patch: `np.ndarray`
+        :param random_location: If True apply patch at randomly shifted locations.
+        :type random_location: `bool`
+        :param channel_index: Index of the color channel.
+        :type channel_index: `Int`
+        """
+        transformations = list()
+        x_copy = x.copy()
+        patch_copy = patch.copy()
+
+        if channel_index == 1:
+            x_copy = np.swapaxes(x_copy, 1, 3)
+            x_copy = np.swapaxes(x_copy, 2, 3)
+            patch_copy = np.swapaxes(patch_copy, 0, 2)
+            patch_copy = np.swapaxes(patch_copy, 1, 2)
+
+        for i_image in range(x.shape[0]):
+
+            if random_location:
+                i_x_1 = random.randint(0, x.shape[1] - 1 - patch.shape[0])
+                i_y_1 = random.randint(0, x.shape[2] - 1 - patch.shape[1])
+            else:
+                i_x_1 = 0
+                i_y_1 = 0
+
+            i_x_2 = i_x_1 + patch.shape[0]
+            i_y_2 = i_y_1 + patch.shape[1]
+
+            transformations.append({"i_x_1": i_x_1, "i_y_1": i_y_1, "i_x_2": i_x_2, "i_y_2": i_y_2})
+
+            x_copy[i_image, i_x_1:i_x_2, i_y_1:i_y_2, :] = patch_copy
```

Comment:
I think the existing code should work because we only update the documentation of the shape convention, it seems to me the indexing was always using the new shape convention.

Reply:
That makes sense :+1:  Agree!